haystack/database/faiss.py

```diff
@@ -0,0 +1,173 @@
+        self.faiss_index: IndexHNSWFlat = self._create_new_index(index_factory=index_factory, vector_size=vector_size)
+        self.index_factory = index_factory
+
+        self.index_buffer_size = index_buffer_size
+        self.phi = 0
+        super().__init__(url=sql_url)
+
+    def _create_new_index(self, index_factory: str, vector_size: int):
+        index = faiss.index_factory(vector_size + 1, index_factory)
+        return index
+
+    def write_documents(self, documents: Union[List[dict], List[Document]], index: Optional[str] = None):
+        if self.phi > 0:
+            raise Exception("Addition of more data in an existing index is not supported.")
+
+        index = index or self.index
+        document_objects = [Document.from_dict(d) if isinstance(d, dict) else d for d in documents]
+
+        add_vectors = False if document_objects[0].embedding is None else True
+
+        if add_vectors:
+            phi = self._get_phi(document_objects)
+            self.phi = 0
+
+        for i in range(0, len(document_objects), self.index_buffer_size):
+            if add_vectors:
+                embeddings = [doc.embedding for doc in document_objects[i : i + self.index_buffer_size]]
+                hnsw_vectors = self._get_hnsw_vectors(embeddings=embeddings, phi=phi)
+                self.faiss_index.add(hnsw_vectors)
+
+            docs_to_write_in_sql = []
+            for vector_id, doc in enumerate(document_objects[i : i + self.index_buffer_size]):
+                meta = doc.meta
+                if add_vectors:
+                    meta["vector_id"] = vector_id
+                docs_to_write_in_sql.append(doc)
+
+            super(FAISSDocumentStore, self).write_documents(docs_to_write_in_sql, index=index)
+
+    def _get_hnsw_vectors(self, embeddings: List[np.array], phi: int) -> np.array:
+        vectors = [np.reshape(emb, (1, -1)) for emb in embeddings]
+        norms = [(doc_vector ** 2).sum() for doc_vector in vectors]
+        aux_dims = [np.sqrt(phi - norm) for norm in norms]
+        hnsw_vectors = [np.hstack((doc_vector, aux_dims[i].reshape(-1, 1))) for i, doc_vector in enumerate(vectors)]
+        hnsw_vectors = np.concatenate(hnsw_vectors, axis=0)
+        return hnsw_vectors
+
+    def _get_phi(self, documents: List[Document]) -> int:
+        phi = 0
+        for doc in documents:
+            norms = (doc.embedding ** 2).sum()  # type: ignore
+            phi = max(phi, norms)
+        return phi
+
+    def update_embeddings(self, retriever: BaseRetriever, index: Optional[str] = None):
+        """
+        Updates the embeddings in the the document store using the encoding model specified in the retriever.
+        This can be useful if want to add or change the embeddings for your documents (e.g. after changing the retriever config).
+
+        :param retriever: Retriever to use to get embeddings for text
+        :param index: Index name to update
+        :return: None
+        """
+        if self.phi > 0:
```

Comment:
This check seems unnecessary, since we create a completely new index in line 117 and then populate it with updated embeddings.